api/src/main/java/jakarta/enterprise/lang/model/AnnotationMemberValue.java

```diff
@@ -2,261 +2,318 @@
      */
     int asInt();
 
     /**
-     * Return the value as a long.
-     * @return The long value
-     * @throws java.lang.IllegalStateException if the value cannot be represented as a long.
+     * Returns the value as a long.
+     *
+     * @return the long value
+     * @throws IllegalStateException if the value cannot be represented as a long.
      */
     long asLong();
 
     /**
-     * Return the value as a float.
-     * @return The float value
-     * @throws java.lang.IllegalStateException if the value cannot be represented as a float.
+     * Returns the value as a float.
+     *
+     * @return the float value
+     * @throws IllegalStateException if the value cannot be represented as a float.
      */
     float asFloat();
 
     /**
-     * Return the value as a double.
-     * @return The double value
-     * @throws java.lang.IllegalStateException if the value cannot be represented as a double.
+     * Returns the value as a double.
+     *
+     * @return the double value
+     * @throws IllegalStateException if the value cannot be represented as a double.
      */
     double asDouble();
 
     /**
-     * Return the value as a char.
-     * @return The char value
-     * @throws java.lang.IllegalStateException if the value cannot be represented as a double.
+     * Returns the value as a char.
+     *
+     * @return the char value
+     * @throws IllegalStateException if this annotation member value is not a char
      */
     char asChar();
 
     /**
-     * Return the value as a string.
-     * @return A string representing the value. Never {@code null}.
+     * Returns the value as a String.
+     *
+     * @return the String value
+     * @throws IllegalStateException if this annotation member value is not a String
      */
     String asString();
 
     /**
-     * Return the value to an enum instance.
-     * @param enumType The enum type
-     * @param <E> The enum generic type
-     * @return The enum instance
-     * @throws java.lang.IllegalStateException if the enum value cannot be established from the given argument
+     * Returns the value as an enum instance.
+     *
+     * @param enumType the enum type
+     * @param <E> the enum generic type
+     * @return the enum instance
+     * @throws IllegalArgumentException if given {@code enumType} is not an enum type
+     * @throws IllegalStateException if this annotation member value is not an enum value
      */
     <E extends Enum<E>> E asEnum(Class<E> enumType);
 
     /**
-     * Return the enum type that an enum value represents.
+     * Returns the enum type of the annotation member value.
      *
-     * @return A {@link jakarta.enterprise.lang.model.declarations.ClassInfo} representing the type of an enum value.
+     * @return a {@link ClassInfo} representing the enum type
+     * @throws IllegalStateException if this annotation member value is not an enum value
      */
     ClassInfo<?> asEnumClass();
```

Comment:
I was wondering if this is necessary since in Micronaut we don't actually store the enum type and only the constant value so it would be complicate matters, given the user can already lookup types via the `Types` impl ?

Reply:
I noticed you dropped this method and I thought it's probably a mistake, so I added it back. What I _didn't_ notice is that you added a parameter to `<E extends Enum<E>> E asEnum(Class<E> enumType)` :-) IMHO, that should just be `<E extends Enum<E>> E asEnum()`. The enum type is available on the declaration of the annotation member, isn't it? As in, you should always be able to find it even if you don't remember it?

Reply:
Looks like as of today we just store enum values as strings. See https://github.com/micronaut-projects/micronaut-core/blob/a949a27c69ff00fe50fa751adf0501540aee79b2/inject-java/src/main/java/io/micronaut/annotation/processing/JavaAnnotationMetadataBuilder.java#L637-L641

I guess we would have to rework that code to store the actual enum values, although I am not sure how possible that is, we would I guess need to store the underlying `VariableElement` from the annotation processing API

Reply:
Alright that sounds weird. I don't know how you [intend to] implement this, but if your implementation of `AnnotationMemberValue` retains a reference to whatever you use to implement `AnnotationInfo`, you should be able to lookup the declaration of the annotation member and from there, lookup the enum type. Am I missing something?

Reply:
yeah probably there is a way, will try work it out

Reply:
OK, thanks. I added a TODO to maybe remove the `Class` parameter from `asEnum`. Let's consider this an open question for now.

Also I realized what was the true nature of your question about `AnnotationMember` / `AnnotationMemberValue`. Currently, `AnnotationMember` is only used on one place: `AnnotationInfo.members()` return `Collection<AnnotationMember>`. That method could easily return `Map<String, AnnotationMemberValue>`, so we could remove `AnnotationMember` altogether. I added a TODO for that as well. (I assume we would then rename `AnnotationMemberValue` to something shorter, perhaps reusing the name `AnnotationMember`; I personally don't like `AnnotationValue`, because that's ambiguous.)

Reply:
So one issue with `asEnum()` with no class parameter at least for annotation processors is that the enum type might not be on the annotation processor classpath and only be on the user class, so it may be impossible to materialize it into an enum value.

I guess in that case we could just throw an exception and say to use `asEnumConstant()` in general though using generic type arguments without a way to infer the type seems odd to me. For example this isn't type safe:

```
Foo foo = member.<Foo>asEnum();
```

And calling asEnum() without specifying the type doesn't compile automatically, or at least IntelliJ doesn't know what to do with it.

Reply:
Oh and +1 to remove `AnnotationMemberValue` and just have `AnnotationMember`

Reply:
Yea I thought `asEnum` might be problematic, I originally had a TODO for that method: `// TODO should this be present?`

Does adding the `Class` parameter help with the classpath issues? If so, I'm fine with leaving it there (and indeed it also help with type inference).

OK, I'll submit another PR to change `AnnotationInfo.members()` to return `Map<String, SomeAnnotationMemberValueType_ProbablyAnnotationMember>` and remove `AnnotationMember` as it currently stands. I've have to figure out one other thing now, but I should get to it later today.

Reply:
Yes having the `Class` parameter is better as the API user knows they have to have the class before they can the enum. That is why I added it in the other PR

Reply:
Here, the API user is the extension author. That is typically not the same person as the extension user.

I'd rather remove `asEnum` completely than having it sometimes failing with classloading errors, to be honest :-)

Reply:
Right, my point was if the extension author has to supply a `Class` parameter, they know that they have to configure their classpath to make sure the `Class` they are requesting is on the same classpath as the extension.

```
FooEnum e = info.asEnum(FooEnum.class);
```

In the above case `FooEnum` will not be referencable if it is not on the classpath of the extension and the extension author would get a compilation error.

Reply:
OK, if you're fine with it, then I am :-) Let's keep `asEnum` as it is now, then. I'll remove the TODO comment.